### PR TITLE
[944] Refactor all database access to use query parameters

### PR DESCRIPTION
**Pull request by f-osorio**
Addresses 944. 

I ran the tests for augur test database using the Docker container. Everything passed. I'm not sure if there's anything else I should have done.

### CODE REVIEW

**Comment by ccarterlandis** [comment]
@sgoggins I don't really know why the docs build is failing, but I'm confident it's not actually related to this PR's content. I think there may have been some issues with RTD themselves, I'll continue to debug though.

As for your other question, I don't really understand why this needs to be tested "under load." These changes aren't going to affect performance in any way, everything should behave exactly the same as before. We are just changing how the text for SQL queries are built.

Going to leave my review of the code in just a minute.

**Comment by gabe-heim** [comment]
@sgoggins I definitely think the target should be changed to a feature branch built off dev and tested before we merge straight to master

File: augur/metrics/commit.py
Comment (ccarterlandis) [inline augur/metrics/commit.py:124]: Please keep the raw SQL, and just change the format of the parameter binding. We use raw SQL statements for every other metric, and having the raw SQL be easily readable is important especially for the metrics in order to understand exactly how they are built, and the format of the data they return.

**Review by gabe-heim — DISMISSED** [review]
Great rewrites especially in the methods like prep_jobs in housekeeper.py, I love this and its something we desperately needed. Thanks so much @f-osorio !! I still think we should test this in a feature branch first though @sgoggins

**Review by ccarterlandis — CHANGES_REQUESTED** [review]
Left a few comments about using raw SQL statements, as well as using `bindparams()`. I only added one comment for `bindparams()`, but it applies in multiple locations.

File: augur/metrics/commit.py
Comment (sgoggins) [reply augur/metrics/commit.py:124]: @ccarterlandis : by "raw sql" do you mean "formatted sql"?  I think this is just formatted poorly, but it is "raw sql". Am I right?

Comment (ccarterlandis) [reply augur/metrics/commit.py:124]: @sgoggins yes, that's what I mean. I chose my words poorly. Besides wanting to remain consistent in our implementation of the metrics, I personally don't like using the `.where` and `.and` methods. AFAIK, they provide no functional benefits over encompassing everything in a `.text` , and they only obscure the actual structure of the query. In our case specifically, we need the metrics to be as easily readable as possible, which is why I requested the change

**Comment by ccarterlandis** [comment]
@f-osorio: I made some comments on this PR a few days ago that were backed by incorrect information and assumptions on my part: my sincere apologies about that! After chatting more with @gabe-heim, I now understand why you opted to use the `s.sql.where` and `s.sql.and_`, since using the query parameters doesn't work with `FROM` statements. 

However, I am still concerned with readability and consistency across the metrics if we use that solution; to that end, I'm going to try to roughly prototype a way to build this query programmatically without using the `s.sql.where/and_` statements that still lets us use parameter binding. I'll report back here on my findings, hopefully with some possible alternative solutions that I'd like to get your input on. Thanks for your hard work on this, we really appreciate it!

**Comment by sgoggins** [comment]
@f-osorio : Do you plan to make the requested changes? Mostly its formatting the SQL so it is readable by other developers. A trial version of Navicat would make that easy.

**Comment by f-osorio** [comment]
@sgoggins I was waiting to see what @ccarterlandis comes up with re: 
> I'm going to try to roughly prototype a way to build this query programmatically without using the s.sql.where/and_ statements that still lets us use parameter binding. I'll report back here on my findings

I'm not totally sure what can be done to improve readability. It's not clear to me in what ways it isn't readable, understanding that it is a different way of presenting a query. 

Short of just these queries to the way they were, the only alternative I can come up with is to create a query for each case and have the `if` statements pull the desired query rather than trying to build the query. The result would be that one query would by replaced by 3 queries (one for each of the tables) with each referring to a different table. Those queries would look like the current form and could use bind parameters, but there's going to be 3 instead of one, which may make maintaining a little more cumbersome.

I'm not sure if that's clear. I can update the pull request with this being implemented in the next few days, if it helps.

**Comment by ccarterlandis** [comment]
@f-osorio looking back at my comments, I was not clear at all in terms of my concerns about "readability"; my sincere apologies for that. My apologies also for not getting back to you about those things I wanted to try, it was rude of me to keep you waiting. 

My main concern is keeping the SQL used to build metric implementations as straightforward and easy to understand as possible; for me, this means avoiding ORM abstractions and using raw SQL statements where possible. I think this helps communicate exactly how the metric is being calculated, which is important for deciding if the result of the metric is one you can trust. The easier you can grok exactly what is being run, the easier it is to understand the machinations of the metric, and how it may or may not be what you're looking for. For consistency's sake, I also prefer raw SQL, simply because that's what every other metric uses. I understand it's not always possible, but I like to avoid introducing special edge cases where possible.

All that being said, I'm getting the feeling that I'm the only person who has (very minor) concerns with the way you've implemented it, and the more I think about it the more I think the changes I originally requested are not really that important. Your changes minimize the sacrifices to readability while keeping a good level of clarity, and add very important security functionality. I don't want to be the one holding up the PR anymore and I want to get this merged for you @f-osorio, so I'll go ahead and approve it.

**Comment by ccarterlandis** [comment]
Well, once I can get these merge conflicts sorted out, I will 😊

**Comment by ccarterlandis** [comment]
I believe I've fixed the merge conflicts in `workers/worker_base.py`, but @gabe-heim & @f-osorio if you are so inclined, it would make me feel better if you would check them as well.

**Review by ccarterlandis — DISMISSED** [review]

**Comment by ccarterlandis** [comment]
Oh no, I totally forgot to come back and merge this. I'll fix the conflicts again and get it merged later today...yikes! Sorry again 😬

**Comment by ccarterlandis** [comment]
This `Travis CI - Branch Expected — Waiting for status to be reported` bug is going to be the death of me! 😡 (not your fault @f-osorio, this has been happening to nearly all of our PRs lately and I haven't figured out why just yet).
Closed and re-opened this PR to try and re-trigger this build so I can merge this

**Comment by sgoggins** [comment]
@f-osorio : Can you push a small change to your branch to trigger our build process? Its a Travis-CI issue we could use your help with.  Thanks!

**Comment by f-osorio** [comment]
I can't tell if it worked. If you need me to do something else, let me know.

**Comment by sgoggins** [comment]
@ccarterlandis : Can you look adn see if this worked?

**Comment by sgoggins** [comment]
This issue has not had comments from the poster in a month, so I am am going to close it @f-osorio and @ccarterlandis ... Happy to reopen it if the work goes forward.

**Pull request closed without merging**
